### PR TITLE
Add marionette

### PR DESCRIPTION
**Pull request by silne30**
Creating pull request for marionette functionality. All tests are now running in series without failures. Will file another bug to add port randomization after that is added to Marionette.

### CODE REVIEW

File: README.rst [hunk 1]
```diff
@@ -25,6 +25,15 @@ Overview
...
 as well as a set of page objects that can be used to interact with Firefox Accounts' sign in screens.
 
+Installation
+------------
+
+``python setup.py develop``
+
+**If running on a Mac, you may need the following before running the above command:**
+
+``pip install cryptography==1.2.1``
```
Comment (davehunt) [inline README.rst]: PyFxA doesn't pin cryptography, so we probably shouldn't either: https://github.com/mozilla/PyFxA/blob/master/setup.py#L39. Let's change this to just `pip install cryptography`

File: .gitignore [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+
+# Crashlytics plugin (for Android Studio and IntelliJ)
+com_crashlytics_export_strings.xml
+crashlytics.properties
+crashlytics-build.properties
+
+# Created by .ignore support plugin (hsz.mobi)
+
+**/*.pyc
+fxapom.egg-info
```
Comment (davehunt) [inline .gitignore]: This file seems excessive. Really I think we just need to ignore:

```
build
dist
*.egg-info
*.pyc
```

I'm fine if you want to include project files though too.

File: README.rst [hunk 3]
```diff
@@ -83,6 +92,12 @@ Running The Tests
 Change Log
 ----------
 
+1.7
+^^^
+
+* Added marionette functionality
+* Added Mac support to Readme
```
Comment (davehunt) [inline README.rst]: This isn't part of the package and so doesn't need to be in the release notes.

File: README.rst [hunk 4]
```diff
@@ -83,6 +92,12 @@ Running The Tests
 Change Log
 ----------
 
+1.7
+^^^
+
+* Added marionette functionality
```
Comment (davehunt) [inline README.rst]: Marionette should be title case and link to the project page.

File: .travis.yml [hunk 5]
```diff
@@ -2,9 +2,27 @@ language: python
 
 python: 2.7
 
-install: "pip install flake8"
+addons:
+  - firefox: "42.0"
```
Comment (davehunt) [inline .travis.yml]: You should be able to use `latest` or `latest-esr` to avoid this needing to be maintained. See https://docs.travis-ci.com/user/firefox/ for details.

File: .travis.yml [hunk 6]
```diff
@@ -2,9 +2,27 @@ language: python
...
+
+before_script:
+- "export DISPLAY=:99.0"
+- "sh -e /etc/init.d/xvfb start"
+- sleep 3 # give xvfb some time to start
+
+install:
+  - sudo apt-get update
+  - sudo apt-get install firefox
```
Comment (davehunt) [inline .travis.yml]: Why is this necessary if we're using the Firefox addon for Travis?

File: .travis.yml [hunk 7]
```diff
@@ -2,9 +2,27 @@ language: python
...
+  - pip install flake8
+  - pip install cryptography
```
Comment (davehunt) [inline .travis.yml]: We shouldn't need to install cryptography separately if we're doing an install rather than a develop.

File: .travis.yml [hunk 8]
```diff
@@ -2,9 +2,27 @@ language: python
...
+  - pip install -r tests/requirements.txt
+
+script:
+  - flake8 .
+  - firefox -v
+  - python setup.py develop
```
Comment (davehunt) [inline .travis.yml]: Why `develop` and not `install`?

File: .travis.yml [hunk 9]
```diff
@@ -2,9 +2,27 @@ language: python
...
+install:
+  - sudo apt-get update
+  - sudo apt-get install firefox
+  - pip install flake8
+  - pip install cryptography
+  - pip install -r tests/requirements.txt
+
+script:
+  - flake8 .
+  - firefox -v
```
Comment (davehunt) [inline .travis.yml]: This information should already be present in the console output for the Firefox addon.

File: .travis.yml [hunk 10]
```diff
@@ -2,9 +2,27 @@ language: python
...
+  - python setup.py develop
+  - py.test --driver=Firefox --capture=sys -s  tests
```
Comment (davehunt) [inline .travis.yml]: `--capture=sys` and `-s` contradict each other as `-s` is a shortcut for `--capture=no`.

File: .gitignore (hunk 2 above)
Comment (davehunt) [inline .gitignore]: There's still a lot of commented out lines here. Can we reduce it to only your project files that you want to exclude?

File: README.rst (hunk 3 above)
Comment (davehunt) [inline README.rst]: Please remove this item.

File: conftest.py [hunk 11]
```diff
@@ -36,3 +29,28 @@ def dev_account():
 @pytest.fixture
 def timeout():
     return TIMEOUT
+
+
+@pytest.fixture
+def marionette(request):
```
Comment (davehunt) [inline conftest.py]: Move this to `tests/marionette/conftest.py`

File: conftest.py [hunk 12]
```diff
@@ -36,3 +29,28 @@ def dev_account():
...
+    m = Marionette(bin=request.config.option.bin)
+    m.start_session()
+    m.set_prefs({'signon.rememberSignons': False})
+    request.addfinalizer(m.delete_session)
+    return m
+
+
+def pytest_addoption(parser):
```
Comment (davehunt) [inline conftest.py]: Move this to `tests/marionette/conftest.py`

File: fxapom/fxapom.py [hunk 13]
```diff
@@ -41,7 +47,6 @@ class FxATestAccount:
 
     def __init__(self, url=DEV_URL):
         """ Creates an FxATestAccount object, which includes a verified account.
-
```
Comment (davehunt) [inline fxapom/fxapom.py]: Please restore this blank line. See http://legacy.python.org/dev/peps/pep-0257/#multi-line-docstrings for details.

File: fxapom/pages/base.py [hunk 14]
```diff
@@ -7,13 +7,12 @@
 
 
 class Base(Page):
-
```
Comment (davehunt) [inline fxapom/pages/base.py]: Please restore this blank line.

File: fxapom/pages/base.py [hunk 15]
```diff
@@ -7,13 +7,12 @@
...
     def __init__(self, selenium, timeout=TIMEOUT):
```
Comment (davehunt) [inline fxapom/pages/base.py]: We should be consistent in renaming `selenium` to `driver`.

File: fxapom/pages/marionette/__init__.py [hunk 16]
```diff
@@ -0,0 +1 @@
+__author__ = 'jdorlus'
```
Comment (davehunt) [inline fxapom/pages/marionette/__init__.py]: Leave this file empty.

File: fxapom/pages/marionette/base.py [hunk 17]
```diff
@@ -0,0 +1,16 @@
+
```
Comment (davehunt) [inline fxapom/pages/marionette/base.py]: Include the license.

File: fxapom/pages/marionette/base.py [hunk 18]
```diff
@@ -0,0 +1,16 @@
...
+from fxapom.fxapom import TIMEOUT
+from page import Page
```
Comment (davehunt) [inline fxapom/pages/marionette/base.py]: We should be consistent with our imports. Try: `from fxapom.pages.page import Page`

File: fxapom/pages/marionette/base.py [hunk 19]
```diff
@@ -0,0 +1,16 @@
...
+
+
+class Base(Page):
+    def __init__(self, driver, timeout=TIMEOUT):
```
Comment (davehunt) [inline fxapom/pages/marionette/base.py]: Insert a blank line before this line.

File: fxapom/pages/marionette/base.py [hunk 20]
```diff
@@ -0,0 +1,16 @@
...
+        super(Page, self).__init__(driver, timeout)
+        self.driver = driver
+        self._main_window_handle = self.driver.current_window_handle
+
+    def switch_to_main_window(self):
+        self.driver.switch_to_window(self._main_window_handle)
+
+    def close_window(self):
+        self.driver.close()
```
Comment (davehunt) [inline fxapom/pages/marionette/base.py]: This file doesn't appear to be any different from `fxapom/pages/base.py`, so it may not even be needed.

File: fxapom/pages/marionette/page.py [hunk 21]
```diff
@@ -0,0 +1,21 @@
...
+class Page(object):
+
+    def __init__(self, driver, timeout=TIMEOUT):
+        self.driver = driver
+        self.timeout = timeout
+
+    def is_element_visible(self, *locator):
+        try:
+            return self.driver.find_element(*locator).is_displayed()
+        except (NoSuchElementException, ):
```
Comment (davehunt) [inline fxapom/pages/marionette/page.py]: Use `except NoSuchElementException:`

File: fxapom/pages/marionette/sign_in.py [hunk 22]
```diff
@@ -0,0 +1,86 @@
...
+    _next_button_locator = (By.ID, 'email-button')
+    _password_input_locator = (By.ID, 'password')
+    _sign_in_locator = (By.ID, 'submit-btn')
+
+    def __init__(self, driver, timeout=TIMEOUT):
+        super(Base, self).__init__(driver, timeout)
+        self._sign_in_window_handle = None
+        self.popup = None
+        self.driver = driver
+        self.timeout = timeout
```
Comment (davehunt) [inline fxapom/pages/marionette/sign_in.py]: These two lines duplicate what's already done by the parent class and so can be removed.

File: fxapom/pages/page.py [hunk 23]
```diff
@@ -2,6 +2,7 @@
 # License, v. 2.0. If a copy of the MPL was not distributed with this file,
 # You can obtain one at http://mozilla.org/MPL/2.0/.
 
+
```
Comment (davehunt) [inline fxapom/pages/page.py]: Remove this extra blank line.

File: fxapom/pages/page.py [hunk 24]
```diff
@@ -9,12 +10,12 @@
...
+    def __init__(self, driver, timeout=TIMEOUT):
+        self.driver = driver
         self.timeout = timeout
 
     def is_element_visible(self, *locator):
         try:
-            return self.selenium.find_element(*locator).is_displayed()
-        except (NoSuchElementException,):
+            return self.driver.find_element(*locator).is_displayed()
+        except (NoSuchElementException, ):
```
Comment (davehunt) [inline fxapom/pages/page.py]: Use `except NoSuchElementException:`

File: fxapom/pages/sign_in.py [hunk 25]
```diff
@@ -2,79 +2,85 @@
...
 # file, You can obtain one at http://mozilla.org/MPL/2.0/.
 
-from base import Base
-
 from selenium.webdriver.common.by import By
 from selenium.webdriver.support import expected_conditions as EC
 from selenium.webdriver.support.ui import WebDriverWait as Wait
 
 from fxapom.fxapom import TIMEOUT
+from base import Base
```
Comment (davehunt) [inline fxapom/pages/sign_in.py]: Use `from fxapom.pages.base import Base`

File: fxapom/pages/sign_in.py [hunk 26]
```diff
@@ -2,79 +2,85 @@
...
 
 
 class SignIn(Base):
 
     _fox_logo_locator = (By.ID, 'fox-logo')
     _email_input_locator = (By.CSS_SELECTOR, '.input-row .email')
+    _marionette_email_input_locator = (By.CSS_SELECTOR, '.input-row .email')
```
Comment (davehunt) [inline fxapom/pages/sign_in.py]: Remove this line.

File: fxapom/pages/sign_in.py [hunk 27]
```diff
@@ -2,79 +2,85 @@
...
     _sign_in_locator = (By.ID, 'submit-btn')
 
-    def __init__(self, selenium, timeout=TIMEOUT):
-        Base.__init__(self, selenium, timeout)
+    def __init__(self, driver, timeout=TIMEOUT):
+        super(Base, self).__init__(driver, timeout)
+        self._sign_in_window_handle = None
+        self.popup = None
+        self.driver = driver
+        self.timeout = timeout
```
Comment (davehunt) [inline fxapom/pages/sign_in.py]: These two lines are not needed.

File: fxapom/pages/sign_in.py [hunk 28]
```diff
@@ -2,79 +2,85 @@
...
+                self.driver.switch_to.window(handle)
+                if self.is_element_visible(*self._fox_logo_locator):
+                    Wait(self.driver, self.timeout).until(
+                        EC.visibility_of_element_located(self._email_input_locator))
+                    self._sign_in_window_handle = handle
+                    break
+            else:
+                raise Exception('Popup has not loaded')
+        else:
+            self.popup = False
```
Comment (davehunt) [inline fxapom/pages/sign_in.py]: Nit: If you set `self.popup` to `False` initially, you wouldn't need these two lines.

File: tests/marionette/__init__.py [hunk 29]
```diff
@@ -0,0 +1 @@
+__author__ = 'jdorlus'
```
Comment (davehunt) [inline tests/marionette/__init__.py]: Leave this file empty.

File: tests/web_driver/__init__.py [hunk 30]
```diff
@@ -0,0 +1 @@
+__author__ = 'jdorlus'
```
Comment (davehunt) [inline tests/web_driver/__init__.py]: Leave this file empty.

File: fxapom/pages/marionette/sign_in.py [hunk 31]
```diff
@@ -0,0 +1,86 @@
...
+
+from fxapom.fxapom import TIMEOUT
+from fxapom.pages.marionette.base import Base
+
+
+class MarionetteSignIn(Base):
+
+    _fox_logo_locator = (By.ID, 'fox-logo')
+    _email_input_locator = (By.CSS_SELECTOR, '.input-row .email')
+    _marionette_email_input_locator = (By.CSS_SELECTOR, '.input-row .email')
```
Comment (davehunt) [inline fxapom/pages/marionette/sign_in.py]: Remove this line.

File: fxapom/pages/marionette/sign_in.py [hunk 32]
```diff
@@ -0,0 +1,86 @@
...
+        self._sign_in_window_handle = None
+        self.popup = None
+        self.driver = driver
+        self.timeout = timeout
+        self.check_for_popup(self.driver.window_handles)
+
+    @property
+    def email(self):
+        """Get the value of the email field."""
+        return self.driver.find_element(*self._email_locator).get_attribute('value')
```
Comment (davehunt) [inline fxapom/pages/marionette/sign_in.py]: This should use the `_email_input_locator`.

File: fxapom/pages/sign_in.py [hunk 33]
```diff
@@ -2,79 +2,85 @@
...
+        password.send_keys(value)
+
+    def click_next(self):
+        self.driver.find_element(*self._next_button_locator).click()
 
     @property
     def email(self):
         """Get the value of the email field."""
-        return self.selenium.find_element(*self._email_input_locator).get_attribute('value')
+        return self.driver.find_element(*self._email_locator).get_attribute('value')
```
Comment (davehunt) [inline fxapom/pages/sign_in.py]: Please revert the locator change, as the new name does not exist.

File: .travis.yml (hunk 10 above)
Comment (davehunt) [inline .travis.yml]: I recommend adding `-r a -v` for more verbose logging in Travis.

**Comment by davehunt** [comment]
@silne30 thanks for following my suggestion of moving the Marionette specific parts into their own modules, it makes the code much more readable! I think the level of duplication we have here is reasonable, and reducing it further may not be worth the effort considering that the Marionette and Selenium APIs are likely to converge in the future. I've added some comments, but once addressed I think we'd be ready for merge and release.

Could you also add a note to the README explaining the Marionette support?

File: .gitignore (hunk 2 above)
Comment (silne30) [inline .gitignore]: Yes, changed. Thanks.

File: README.rst (hunk 3 above)
Comment (silne30) [inline README.rst]: Removed.

File: README.rst (hunk 4 above)
Comment (silne30) [inline README.rst]: Done. Linked to ReadTheDocs.

File: .travis.yml (hunk 5 above)
Comment (silne30) [inline .travis.yml]: Got it.

File: .travis.yml (hunk 6 above)
Comment (silne30) [inline .travis.yml]: We spoke in channel about this. Fixed.

File: .travis.yml (hunk 7 above)
Comment (silne30) [inline .travis.yml]: Changed.

File: .travis.yml (hunk 8 above)
Comment (silne30) [inline .travis.yml]: Because I am a newb. :D Changed.

File: .travis.yml (hunk 9 above)
Comment (silne30) [inline .travis.yml]: Removed.

File: .travis.yml (hunk 10 above)
Comment (silne30) [inline .travis.yml]: Fixed.

File: conftest.py (hunk 11 above)
Comment (silne30) [inline conftest.py]: Done.

File: conftest.py (hunk 12 above)
Comment (silne30) [inline conftest.py]: Done.

File: fxapom/pages/base.py (hunk 14 above)
Comment (silne30) [inline fxapom/pages/base.py]: Restored.

File: fxapom/fxapom.py (hunk 13 above)
Comment (silne30) [inline fxapom/fxapom.py]: Restored. I did not know that.

File: fxapom/pages/base.py (hunk 15 above)
Comment (silne30) [inline fxapom/pages/base.py]: Fixed.

File: fxapom/pages/marionette/__init__.py (hunk 16 above)
Comment (silne30) [inline fxapom/pages/marionette/__init__.py]: Fixed.

File: fxapom/pages/marionette/base.py (hunk 17 above)
Comment (silne30) [inline fxapom/pages/marionette/base.py]: Done.

File: fxapom/pages/marionette/base.py (hunk 18 above)
Comment (silne30) [inline fxapom/pages/marionette/base.py]: Done. 
used from fxapom.pages.marionette.page import Page

File: fxapom/pages/marionette/base.py (hunk 19 above)
Comment (silne30) [inline fxapom/pages/marionette/base.py]: Done.

File: fxapom/pages/marionette/page.py (hunk 21 above)
Comment (silne30) [inline fxapom/pages/marionette/page.py]: Done.

File: fxapom/pages/marionette/base.py (hunk 20 above)
Comment (silne30) [inline fxapom/pages/marionette/base.py]: I added it because you mentioned that I should likely leave the original page.py as is which has functionality in base.py that I can't change. Because of that, I just rewrote a marionette version of base and page.py

File: fxapom/pages/page.py (hunk 23 above)
Comment (silne30) [inline fxapom/pages/page.py]: Done.

File: fxapom/pages/page.py (hunk 24 above)
Comment (silne30) [inline fxapom/pages/page.py]: OK. This was how it was originally written. Changed.

File: fxapom/pages/sign_in.py (hunk 25 above)
Comment (silne30) [inline fxapom/pages/sign_in.py]: Done.

File: fxapom/pages/sign_in.py (hunk 26 above)
Comment (silne30) [inline fxapom/pages/sign_in.py]: Done.

File: fxapom/pages/sign_in.py (hunk 27 above)
Comment (silne30) [inline fxapom/pages/sign_in.py]: Done.

File: fxapom/pages/sign_in.py (hunk 28 above)
Comment (silne30) [inline fxapom/pages/sign_in.py]: Nice catch.

File: tests/marionette/__init__.py (hunk 29 above)
Comment (silne30) [inline tests/marionette/__init__.py]: Done.

File: tests/web_driver/__init__.py (hunk 30 above)
Comment (silne30) [inline tests/web_driver/__init__.py]: Done.

File: fxapom/pages/marionette/sign_in.py (hunk 31 above)
Comment (silne30) [inline fxapom/pages/marionette/sign_in.py]: Done.

File: fxapom/pages/marionette/sign_in.py (hunk 32 above)
Comment (silne30) [inline fxapom/pages/marionette/sign_in.py]: Done.

File: README.rst (hunk 4 above)
Comment (davehunt) [inline README.rst]: I think https://developer.mozilla.org/en-US/docs/Mozilla/QA/Marionette might be a better place to link to.

File: fxapom/pages/marionette/base.py (hunk 20 above)
Comment (davehunt) [inline fxapom/pages/marionette/base.py]: Okay, the the important factor here is the parent class for `Base`? Makes sense then.

File: fxapom/pages/page.py (hunk 24 above)
Comment (davehunt) [inline fxapom/pages/page.py]: Yeah, I noticed that. Thanks for cleaning it up.

File: fxapom/pages/sign_in.py (hunk 33 above)
Comment (silne30) [inline fxapom/pages/sign_in.py]: Done.

File: fxapom/pages/marionette/sign_in.py (hunk 32 above)
Comment (silne30) [inline fxapom/pages/marionette/sign_in.py]: Done.

File: README.rst (hunk 4 above)
Comment (silne30) [inline README.rst]: Sounds like a plan. Thanks.

**Comment by silne30** [comment]
@davehunt I believe that this is ready.

File: .travis.yml [hunk 34]
```diff
@@ -2,9 +2,23 @@ language: python
...
+- sleep 3 # give xvfb some time to start
+
+install:
+  - pip install flake8
+  - pip install -r tests/requirements.txt
+
+script:
+  - flake8 .
+  - python setup.py install
+  - py.test --driver=Firefox --capture=sys -r a -v  tests
```
Comment (davehunt) [inline .travis.yml]: Why do you set `--capture=sys`?

File: README.rst [hunk 35]
```diff
@@ -38,6 +47,10 @@ or the production environment, which are:
 * ``fxapom.DEV_URL`` - the url for the development environment
 * ``fxapom.PROD_URL`` - the url for the production environment
 
+FxA POM is now able to handle tests written using both Selenium WebDriver and Marionette.
```
Comment (davehunt) [inline README.rst]: Nit: FxAPOM

File: conftest.py [hunk 36]
```diff
@@ -36,3 +28,19 @@ def dev_account():
 @pytest.fixture
 def timeout():
     return TIMEOUT
+
+
+def pytest_addoption(parser):
```
Comment (davehunt) [inline conftest.py]: This still needs to be moved to `tests/marionette` as it's not used by Selenium.

File: fxapom/pages/marionette/__init__.py (hunk 16 above)
Comment (davehunt) [inline fxapom/pages/marionette/__init__.py]: Did you forget to push this fix?

File: fxapom/pages/marionette/page.py [hunk 37]
```diff
@@ -0,0 +1,21 @@
+# This Source Code Form is subject to the terms of the Mozilla Public
+# License, v. 2.0. If a copy of the MPL was not distributed with this file,
+# You can obtain one at http://mozilla.org/MPL/2.0/.
+
+
```
Comment (davehunt) [inline fxapom/pages/marionette/page.py]: Nit: Remove one of these blank lines.

File: fxapom/pages/marionette/sign_in.py (hunk 22 above)
Comment (davehunt) [inline fxapom/pages/marionette/sign_in.py]: This still needs to be addressed.

File: fxapom/pages/marionette/sign_in.py [hunk 38]
```diff
@@ -0,0 +1,85 @@
...
+                self.driver.switch_to.window(handle)
+                if self.is_element_visible(*self._fox_logo_locator):
+                    Wait(self.driver, self.timeout).until(
+                        expected.element_displayed(*self._email_input_locator))
+                    self._sign_in_window_handle = handle
+                    break
+            else:
+                raise Exception('Popup has not loaded')
+        else:
+            self.popup = False
```
Comment (davehunt) [inline fxapom/pages/marionette/sign_in.py]: For consistency, we should initially set this to `False` and remove these two lines.

File: fxapom/pages/sign_in.py [hunk 39]
```diff
@@ -19,62 +18,65 @@ class SignIn(Base):
     _password_input_locator = (By.ID, 'password')
     _sign_in_locator = (By.ID, 'submit-btn')
 
-    def __init__(self, selenium, timeout=TIMEOUT):
-        Base.__init__(self, selenium, timeout)
+    def __init__(self, driver, timeout=TIMEOUT):
+        super(Base, self).__init__(driver, timeout)
+        self._sign_in_window_handle = None
+        self.popup = None
```
Comment (davehunt) [inline fxapom/pages/sign_in.py]: This should be `False` initially.

File: fxapom/pages/marionette/sign_in.py (hunk 22 above)
Comment (silne30) [inline fxapom/pages/marionette/sign_in.py]: Whoops. Done.

File: .travis.yml (hunk 34 above)
Comment (silne30) [inline .travis.yml]: Removed.

File: README.rst (hunk 35 above)
Comment (silne30) [inline README.rst]: Fair enough. Fixed.

File: conftest.py (hunk 36 above)
Comment (silne30) [inline conftest.py]: Done.

File: fxapom/pages/marionette/page.py (hunk 37 above)
Comment (silne30) [inline fxapom/pages/marionette/page.py]: Why doesn't flake8 catch this stuff?

File: fxapom/pages/marionette/sign_in.py (hunk 38 above)
Comment (silne30) [inline fxapom/pages/marionette/sign_in.py]: Done.

File: fxapom/pages/sign_in.py (hunk 39 above)
Comment (silne30) [inline fxapom/pages/sign_in.py]: Ah, good catch.

File: fxapom/pages/marionette/page.py (hunk 37 above)
Comment (davehunt) [inline fxapom/pages/marionette/page.py]: This isn't strictly a pep8 issue.

File: fxapom/pages/base.py [hunk 40]
```diff
@@ -8,12 +8,13 @@
 
 class Base(Page):
 
-    def __init__(self, selenium, timeout=TIMEOUT):
-        super(Base, self).__init__(selenium, timeout)
-        self._main_window_handle = self.selenium.current_window_handle
+    def __init__(self, driver, timeout=TIMEOUT):
+        self.driver = driver
```
Comment (davehunt) [inline fxapom/pages/base.py]: Nit: This is not needed as it's done in `Page.__init__`

File: fxapom/pages/marionette/base.py [hunk 41]
```diff
@@ -0,0 +1,21 @@
...
+from fxapom.fxapom import TIMEOUT
+from fxapom.pages.marionette.page import Page
+
+
+class Base(Page):
+
+    def __init__(self, driver, timeout=TIMEOUT):
+        super(Page, self).__init__(driver, timeout)
+        self.driver = driver
+        self.timeout = timeout
```
Comment (davehunt) [inline fxapom/pages/marionette/base.py]: Nit: These two lines are not needed as they're also done in `Page.__init__`

File: tests/web_driver/test_fxa_login.py [hunk 42]
```diff
@@ -13,7 +13,7 @@ class TestLogin(object):
 
     _fxa_logged_in_indicator_locator = (By.ID, 'loggedin')
 
-    def test_user_can_sign_in(self, selenium, dev_account, click_login, timeout):
+    def test_user_can_sign_in(self, base_url, selenium, dev_account, click_login, timeout):
```
Comment (davehunt) [inline tests/web_driver/test_fxa_login.py]: Nit: We don't need the `base_url` fixture here.

**Comment by davehunt** [comment]
Nit: Please rename `tests/web_driver` to `tests/webdriver`

**Comment by davehunt** [comment]
Just nits, once these are addressed I'd be happy to merge and push a new release!

File: fxapom/pages/base.py (hunk 40 above)
Comment (silne30) [inline fxapom/pages/base.py]: Done.

File: fxapom/pages/marionette/base.py (hunk 41 above)
Comment (silne30) [inline fxapom/pages/marionette/base.py]: Done

File: tests/web_driver/test_fxa_login.py (hunk 42 above)
Comment (silne30) [inline tests/web_driver/test_fxa_login.py]: Done.

**Comment by silne30** [comment]
I have never pushed anything that was on PyPi before. This is pretty exciting.

**Comment by davehunt** [comment]
@silne30: In that case, [sign up for an account](https://pypi.python.org/pypi?%3Aaction=register_form) and let me know your username. I'll add you as a maintainer, and you can do the push! :smile:

**Comment by davehunt** [comment]
Merged in https://github.com/mozilla/fxapom/commit/76c67e78870cd1cb463475a7832475e4e2522918
